### PR TITLE
Add feature for desktop users, sending message by Enter button

### PR DESCRIPTION
**Pull request by YaroslavOvdii**
@tonytlwu @squallstar 
## Issue
https://github.com/Fliplet/fliplet-studio/issues/3889

## Description
Fixed queue and added for desktop users possibility to send a message by press Enter button 

## Screenshots/screencasts
08a03695-52e6-4a2e-9ebb-61b02b5164ff-video

## Backward compatibility

This change is fully backward compatible.

### CODE REVIEW

File: js/queue.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+var Queue = function () {
+  this.messageQueue = [];
+};
+
+Queue.prototype.add = function (message) {
```
Comment (squallstar) [inline js/queue.js]: Rename to "Queue.push"

File: js/queue.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+Queue.prototype.init = function (messages) {
+  this.messageQueue = messages;
+};
+
+Queue.prototype.getAllQueue = function () {
+  return this.messageQueue;
+};
+
+Queue.prototype.sended = function (messages) {
```
Comment (squallstar) [inline js/queue.js]: What does this method return? p.s. the past participle of "send" is "sent"

File: js/queue.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      if (message.guid === sendedItem.guid) {
+        message.sended = true;
+      }
+    });
+  });
+
+  return this.messageQueue;
+};
+
+Queue.prototype.getNotSended = function () {
```
Comment (squallstar) [inline js/queue.js]: getUnsent

File: js/queue.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  return _.filter(this.messageQueue, ['sended', false]);
+};
+
+Queue.prototype.remove = function (message) {
```
Comment (squallstar) [inline js/queue.js]: Queue.pull

File: js/queue.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+var Queue = function () {
```
Comment (squallstar) [inline js/queue.js]: Queue is a very generic variable name, let's rename it to ChatMessagesQueue

File: js/queue.js (hunk 2 above)
Comment (YaroslavOvdii) [reply js/queue.js]: @squallstar that's my mistake it shouldn't return anything. This method serves only to label messages that we have sent already, so we won't send them again. I'll remove the return statement here.

File: js/build.js [hunk 6]
```diff
@@ -1555,6 +1481,113 @@ Fliplet.Widget.instance('chat', function (data) {
...
+  function processMessage() {
+    var unsentMessages = queue.getUnsent();
+    var sendReqPromises = [];
+
+    if (Fliplet.Navigator.isOnline()) {
+      unsentMessages.forEach(function (message) {
+        sendReqPromises.push(chat.message(currentConversation.id, message));
+      });
+
+      queue.sended(unsentMessages);
```
Comment (squallstar) [inline js/build.js]: `sended` doesn't seem to be defined in the queue anymore. Did you actually test this code?

File: js/build.js [hunk 7]
```diff
@@ -2330,9 +2363,9 @@ Fliplet.Widget.instance('chat', function (data) {
       }
     });
 
-    if (messagesQueue.length) {
-      setTimeout(function() {
-        messagesQueue.forEach(function(message) {
+    if (queue.getAllQueue().length) {
```
Comment (squallstar) [inline js/build.js]: I would rename `getAllQueue()` to `getMessages()`

File: js/build.js [hunk 8]
```diff
@@ -1555,6 +1481,113 @@ Fliplet.Widget.instance('chat', function (data) {
...
+
+  function changeUIOnMessageSent($parentHolder, $holder) {
+    // UI CHANGES
+    $messageArea.val('');
+    autosize.update($messageArea);
+    $parentHolder.removeClass('ready');
+    $holder.removeClass('sending');
+    resetImages();
+
+    // RETURN
```
Comment (squallstar) [inline js/build.js]: let's get rid of this useless comment

Comment (YaroslavOvdii) [reply js/build.js]: @squallstar Fixed

File: js/build.js (hunk 7 above)
Comment (YaroslavOvdii) [reply js/build.js]: @squallstar Fixed

File: js/build.js (hunk 6 above)
Comment (YaroslavOvdii) [reply js/build.js]: @squallstar Fixed

File: js/queue.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ChatMessagesQueue.prototype.getMessages = function () {
+  return this.messageQueue;
+};
+
+ChatMessagesQueue.prototype.sent = function (messages) {
+  var messageQueue = this.messageQueue;
+  messages.forEach(function (sentItem) {
+    messageQueue.forEach(function (message) {
+      if (message.guid === sentItem.guid) {
+        message.sent = true;
```
Comment (squallstar) [inline js/queue.js]: Aren't messages being removed from the queue after being sent instead of being marked as sent?

Comment (YaroslavOvdii) [reply js/queue.js]: @squallstar no, they are in processing by `chat.message()` function. Which encodes them and save them to the data source. After what sending saved message into this function `onNewMessage()`. Where we remove saved a message from the queue with this method `queue.pull()` and update `Fliplet.Storage` with an updated queue.
That is how it was designed from the start. I just split this order from one large function to smaller ones.

**Pull request merged**